presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/test/ConnectorRuleTester.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule.test;
+
+import io.prestosql.spi.connector.ConnectorFactory;
+import io.prestosql.sql.parser.SqlParser;
+import io.prestosql.sql.planner.TypeAnalyzer;
+import io.prestosql.sql.planner.iterative.rule.PushLimitIntoTableScan;
+import io.prestosql.sql.planner.iterative.rule.PushPredicateIntoTableScan;
+import io.prestosql.sql.planner.iterative.rule.PushProjectionIntoTableScan;
+import io.prestosql.sql.planner.iterative.rule.PushSampleIntoTableScan;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+
+import java.util.Map;
+
+import static io.airlift.testing.Closeables.closeAllRuntimeException;
+
+public abstract class ConnectorRuleTester
```

Comment:
I am not sure of the value of it. I think it was deliberate decision to not extract things like that. To avoid `io.prestosql.tests.AbstractTestQueryFramework` burden.

Reply:
@kokosing Can you please elaborate on what you mean about `AbstractTestQueryFramework`? The `ConnectorRuleTester` ends up using a LocalQueryRunner.

With new optimizers like `PushXIntoTableScan`, we have started relying on connectors to provide new handles for certain connector expressions. But, the space of expressions that are (1) "translatable" to the connector expressions, and (2) "pushdown-able" in different connectors - will keep evolving independently. (For example, say we start translating lambda expressions to ConnectorExpressions, but the Hive Connector is not ready to support it.)  

I feel that tests that cover plugging in `PushXIntoTableScan` and `ConnectorMetadata::applyX` are required to be confident about the compatibility of these changes going forward, with all connectors. And it will be useful to have them at a more granular level than just end-to-end test, to make sure that we're not producing suboptimal plans for any connector, while adding new features. What do you think?

Reply:
My point was to have something that we could use reuse using composition like basic `RuleTester` and `QueryAssertions`. I would prefer to avoid inheritance if possible here. 

`AbstractTestQueryFramework` is a root class of many test classes. When you want to do something custom it requires to pass parameters through many constructors and provide a abstract methods that are implemented in particualar moment and implemented only somewhere. In short it is quite difficult to maintain.

Having possibility to use composition would allow us to use something like junit `@Rule` at some point.